Hw21 cache

SGrechnev opened the pull request:
run DB
`hw21-cache/docker/runDB.cmd`
run demo with correct VM options (-Xmx16m -Xms16m)
`gradlew run`

DEMO SPEED
 - DbServiceClient with MyCache is faster

DEMO FREE
 - MyCache with HashMap results in OOM
 - MyCache with WeakHashMap lets you avoids OOM

SGrechnev committed:
Implement MyCache
Listeners in List<WeakReference<HwListener>>
!!! ATTENTION !!!
In fact listener2 is removed on line 70 (in for-loop)

SGrechnev committed:
Implement DbServiceClient with Cache
'hw21-cache/docker/runDB.cmd' to run DB
'gradlew run' for run demo with correct VM options (-Xmx16m -Xms16m)
DEMO SPEED
 - DbServiceClient with MyCache is faster
DEMO FREE
 - MyCache with HashMap results in OOM
 - MyCache with WeakHashMap lets you avoids OOM

petrelevich commented on hw21-cache/src/main/java/ru/otus/cachehw/MyCache.java:
```diff
@@ -0,0 +1,88 @@
+            // Элемент по ключу не найден
+            return;
+        }
+        innerMap.remove(key);
```
обратите внимание, что возвращает remove:

> Returns:
> the previous value associated with key, or null if there was no mapping for key.

petrelevich commented on hw21-cache/src/main/java/ru/otus/cachehw/MyCache.java:
```diff
@@ -0,0 +1,88 @@
+    @Override
+    public void removeListener(HwListener<K, V> listener) {
+        var weakListenersToRemove = listeners.stream()
+                .filter(l -> l.get()==listener)
```
почему не equals ?

petrelevich commented on hw21-cache/src/main/java/ru/otus/cachehw/MyCache.java:
```diff
@@ -0,0 +1,88 @@
+        for(var iterator = listeners.iterator(); iterator.hasNext();){
+            var listener = iterator.next().get();
+            if (listener != null) {
+                listener.notify(key, value, action);
```
notify лучше в try-catch вызывать.

petrelevich commented on hw21-cache/src/main/java/ru/otus/cachehw/MyCache.java:
```diff
@@ -0,0 +1,88 @@
+        return listeners.stream().reduce(
+                false,
+                (acc, l) -> acc || (l.get().equals(listener)),
+                (acc, bool) -> acc || bool);
```
кажется, в цикле перебирать намного проще, а если значение нашлось, можно сразу return true сделать.

petrelevich commented on hw21-cache/src/main/java/ru/otus/core/service/DBServiceClient.java:
```diff
@@ -0,0 +1,22 @@
+
+    void setUseCache(boolean useCache);
+
+    MyCache<Long, Client> getClientCache();
```
пересмотрите пример из вебинара про использование Long и Integer в качестве ключей.

petrelevich commented on hw21-cache/src/main/java/ru/otus/core/service/DbServiceClientImpl.java:
```diff
@@ -0,0 +1,74 @@
+        try (SessionManager sessionManager = clientDao.getSessionManager()) {
+            sessionManager.beginSession();
+            try {
+                return clientDao.findById(id);
```
если значение не нашлось в кеше, то стоит его туда положить.

petrelevich commented on hw21-cache/src/main/java/ru/otus/core/service/DbServiceClientImpl.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @Override
+    public MyCache<Long, Client> getClientCache() {
```
сомнительный метод в интерфейсе DBServiceClient, тем более, что к кешу можно при желании по ссылке добраться.

petrelevich commented on hw21-cache/src/main/java/ru/otus/core/service/DbServiceClientImpl.java:
```diff
@@ -0,0 +1,74 @@
+    private boolean useCache;
+
+    public boolean isUseCache() {
+        return useCache;
```
лучше вместо такого поля сделать имплементацию DBServiceClient с кешом.
на крайний случай передавать useCache в конструкторе.

SGrechnev committed:
Refactoring
 - split DbServiceClientImpl into implementations with and without cache
 - MyCache uses String as a key
